planner/plansdk/plansdk.go

```diff
@@ -122,32 +122,51 @@ func MergePlans(plans ...*Plan) *Plan {
 	plan := &Plan{
 		DevPackages:     []string{},
 		RuntimePackages: []string{},
-		SharedPlan: SharedPlan{
-			InstallStage: &Stage{},
-			BuildStage:   &Stage{},
-			StartStage:   &Stage{},
-		},
 	}
 	for _, p := range plans {
-		err := mergo.Merge(plan, p, mergo.WithAppendSlice)
+		err := mergo.Merge(
+			plan,
+			&Plan{
+				DevPackages:     p.DevPackages,
+				RuntimePackages: p.RuntimePackages,
+			},
+			// Only WithAppendSlice the dev and runtime packages field.
+			mergo.WithAppendSlice,
+		)
 		if err != nil {
 			panic(err) // TODO: propagate error.
 		}
 	}
 
 	plan.DevPackages = pkgslice.Unique(plan.DevPackages)
 	plan.RuntimePackages = pkgslice.Unique(plan.RuntimePackages)
+	plan.SharedPlan = PickSharedPlan(plans...)
 
-	// Set default files for install stage to copy.
-	if plan.SharedPlan.InstallStage.InputFiles == nil {
-		plan.SharedPlan.InstallStage.InputFiles = []string{"."}
+	return plan
+}
+
+func PickSharedPlan(plans ...*Plan) SharedPlan {
+	for _, p := range plans {
+		// For now, pick the first buildable plan.
+		if p.Buildable() {
+			return p.SharedPlan
+		}
 	}
-	// Set default files for install stage to copy over from build step.
-	if plan.SharedPlan.StartStage.InputFiles == nil {
-		plan.SharedPlan.StartStage.InputFiles = []string{"."}
+	return SharedPlan{}
+}
+
+func OverrideSharedPlan(dst *Plan, src *Plan) SharedPlan {
+	sharedPlan := &Plan{
+		SharedPlan: dst.SharedPlan,
+	}
+	// fields in the dst plan:
+	//   if empty, will inherit the corresponding src fields
+	//   if set, will override corresponding src fields
+	if err := mergo.Merge(sharedPlan, src); err != nil {
+		panic(err) // TODO: propagate error.
```

Comment:
nit, let's avoid panics and instead structure the code to return errors.

Reply:
The `panic` is everywhere at this point. Maybe we need a cleanup round to consolidate them all.

Reply:
I think we should avoid adding new ones. (each new one makes the task of getting rid of them harder).

This function is only called by `devbox.Plan` which itself is only called by 3 functions (all of which return error). so it feels like a pretty minor change.